Source/WebCore/page/scrolling/ScrollAnchoringController.cpp

```diff
@@ -50,35 +53,138 @@ ScrollOffset ScrollAnchoringController::computeOffset(RenderBox& candidate)
+        // TODO: figure out how to get scrollable area for renderer to check if it is maintaining scroll anchor
+        if (renderer->style().overflowAnchor() == OverflowAnchor::None || renderer->isStickilyPositioned() || renderer->isFixedPositioned() || renderer->isPseudoElement() || renderer->isAnonymousBlock() || (renderer->isInline() && !renderer->isAtomicInlineLevelBox()))
+            return CandidateExaminationResult::Exclude;
+        if (&element == document->bodyOrFrameset() || is<HTMLHtmlElement>(&element))
+            return CandidateExaminationResult::Skip;
+        auto boxRect = renderer->absoluteBoundingBoxRect();
+        if (!boxRect.width() || !boxRect.height())
+            return CandidateExaminationResult::Skip;
+        if (layoutViewport.contains(boxRect))
+            return CandidateExaminationResult::Select;
+        if (layoutViewport.intersects(boxRect))
+            return CandidateExaminationResult::Descend;
+    }
+    return CandidateExaminationResult::Skip;
+}
+#if !LOG_DISABLED
+static TextStream& operator<<(TextStream& ts, CandidateExaminationResult result)
+{
+    switch (result) {
+    case CandidateExaminationResult::Exclude:
+        ts << "Exclude";
+        break;
+    case CandidateExaminationResult::Select:
+        ts << "Select";
+        break;
+    case CandidateExaminationResult::Descend:
+        ts << "Descend";
+        break;
+    case CandidateExaminationResult::Skip:
+        ts << "Skip";
+        break;
+    }
+    return ts;
+}
+#endif
+
+Element* ScrollAnchoringController::findAnchorElementRecursive(Element* element)
+{
+    if (!element)
+        return nullptr;
+
+    auto result = examineCandidate(*element);
+    LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::findAnchorElementRecursive() element: "<< *element<<" examination result: " << result);
+
+    switch (result) {
+    case CandidateExaminationResult::Select:
+        return element;
+    case CandidateExaminationResult::Exclude:
+        return nullptr;
+    case CandidateExaminationResult::Skip:
+    case CandidateExaminationResult::Descend: {
+        for (auto& child : childrenOfType<Element>(*element)) {
+            if (auto* anchorElement = findAnchorElementRecursive(&child))
+                return anchorElement;
         }
     }
-    if (!candidateAnchor)
+    }
+    if (result == CandidateExaminationResult::Skip)
+        return nullptr;
+    return element;
+}
+
+void ScrollAnchoringController::chooseAnchorElement(Document& document)
+{
+    LOG_WITH_STREAM(ScrollAnchoring, stream << "ScrollAnchoringController::chooseAnchorElement() starting findAnchorElementRecursive: ");
+
+    findPriorityCandidate(document);
+
+    Element* anchorElement = nullptr;
+
+    // TODO: iterate from owning scroller not root
+    if (!m_anchorElement)
+        anchorElement = findAnchorElementRecursive(document.documentElement());
+
+    if (!anchorElement)
```

Comment:
A bit confused about why we don't change `m_anchorElement` if we fail to find a new one.

Reply:
We only call `chooseAnchorElement` if `m_anchorElement` is null.

Reply:
I feel like nesting the ifs would make this more clear.

Reply:
Fixed.